b2sdk/utils/curl.py

```diff
@@ -0,0 +1,431 @@
+        # for `run_iteration` it's possible that your curl already has something to show.
+        # This is a separate lock than the perform lock, as the latter has to be re-entrant.
+        if not self.run_lock.acquire(timeout=self.ACQUIRE_SLEEP_SECONDS):
+            return
+        try:
+            self._run_iteration()
+        finally:
+            self.run_lock.release()
+
+    def _run_iteration(self) -> None:
+        # We could use select here for cases when there is nothing to do (no curls registered)
+        # but this provides us with no actual gain. `run_iteration` is called by any thread
+        # that wants data, so naturally there are operations awaiting data.
+        with self.perform_lock:
+            # Perform also have to be inside this lock to ensure that neither `add_handler`
+            # nor `Curl.getinfo` is called during this operation.
+            status, _count = self.multi.perform()
+            # Usual culprits here could be "out of memory" and "internal error",
+            # we can't really do anything about either of them.
+            assert status == pycurl.E_MULTI_OK, status
+
+            # Check whether any of the curls finished, in one way or another.
+            remaining_messages_count, successful_curls, failed_curls = self.multi.info_read()
+            # Remove each finished curl from our multi object and close it.
+            finished_curls = successful_curls + [elem[0] for elem in failed_curls]
+            # TODO: additional handling for failed curls.
+            for curl in finished_curls:
+                self.multi.remove_handle(curl)
+                streamer = self.mapping[curl]
+                streamer.close()
+
+
+@dataclass
+class CurlResponse:
+    """
+    Analog of requests.Response
+
+    Provides basic interface for interaction with HTTP.
+
+    Note that in streaming mode it can be completely "empty",
+    that is neither status code nor headers are downloaded initially.
+    This means that first call to `status_code`, `headers` or
+    `iter_content` (through e.g. `content`) will start the actual operations.
+    """
+
+    streamer: CurlStreamer
+    request: CurlRequest
+
+    content_cache: Optional[bytes] = None
+
+    @property
+    def url(self) -> str:
+        return self.request.url
+
+    @property
+    def status_code(self) -> int:
+        return self.streamer.status_code
+
+    @property
+    def headers(self) -> CaseInsensitiveDict:
+        return self.streamer.headers
+
+    @property
+    def ok(self) -> bool:
+        return 200 <= self.status_code < 400
+
+    @property
+    def content(self) -> bytes:
+        if self.content_cache is None:
+            self.content_cache = b''.join(self.iter_content(1024))
+        return self.content_cache
+
+    def iter_content(self, chunk_size: int) -> Iterator[bytes]:
+        while True:
+            data = self.streamer.read(chunk_size)
```

Comment:
shouldn't `read` block if there is no data yet available but it's not the end of the stream?

Reply:
I should rename this, because it does `read` from a stream, but behaves slightly differently to what we're used to. And it's because I wanted to not have another thread lingering in the background.